### PR TITLE
hash: add yield version of lookup function

### PR DESCRIPTION
**Pull request by mengxiang0811**

### CODE REVIEW

File: lib/librte_hash/rte_cuckoo_hash.c [hunk 1]
```diff
@@ -1565,6 +1774,146 @@ rte_hash_del_key_with_hash(const struct rte_hash *h,
...
+
+	short_sig = get_short_sig(sig);
+	prim_bucket_idx = get_prim_bucket_index(h, sig);
+	sec_bucket_idx = get_alt_bucket_index(h, prim_bucket_idx, short_sig);
+	prim_bkt = &h->buckets[prim_bucket_idx];
+	yield_func(prim_bkt, arg);
+
+	__hash_rw_writer_lock(h);
+	/* look for key in primary bucket */
+	ret = search_and_remove(h, key, prim_bkt, short_sig, &pos);
```
Comment (AltraMayor) [inline lib/librte_hash/rte_cuckoo_hash.c]: Why are we not using `search_and_remove_and_yield()` here?

File: lib/librte_hash/rte_cuckoo_hash.c [hunk 2]
```diff
@@ -1565,6 +1774,146 @@ rte_hash_del_key_with_hash(const struct rte_hash *h,
...
+		prev_bkt = prim_bkt;
+		goto return_bkt;
+	}
+
+	/* Calculate secondary hash */
+	sec_bkt = &h->buckets[sec_bucket_idx];
+
+	FOR_EACH_BUCKET(cur_bkt, sec_bkt) {
+		yield_func(cur_bkt, arg);
+		ret = search_and_remove(h, key, cur_bkt, short_sig, &pos);
```
Comment (AltraMayor) [inline lib/librte_hash/rte_cuckoo_hash.c]: Why are we using `search_and_remove_and_yield()` here?

File: lib/librte_hash/rte_cuckoo_hash.c [hunk 3]
```diff
@@ -1565,6 +1774,146 @@ rte_hash_del_key_with_hash(const struct rte_hash *h,
...
+	if (!last_bkt) {
+		__hash_rw_writer_unlock(h);
+		return ret;
+	}
+	while (last_bkt->next) {
+		prev_bkt = last_bkt;
+		last_bkt = last_bkt->next;
+	}
+
+	yield_func(last_bkt, arg);
```
Comment (AltraMayor) [inline lib/librte_hash/rte_cuckoo_hash.c]: This function should come before the `while` loop above AND as the last line of the `while` loop.

**Review by AltraMayor — CHANGES_REQUESTED** [review]
**hash: add yield version of lookup and deletion functions**

File: lib/librte_hash/rte_cuckoo_hash.c (hunk 2 above)
Comment (mengxiang0811) [reply lib/librte_hash/rte_cuckoo_hash.c]: Shouldn't we use ``search_and_remove_and_yield()``? As it also needs to go over the bucket and finds the key to be removed.

Comment (AltraMayor) [reply lib/librte_hash/rte_cuckoo_hash.c]: I meant to write "Why are we NOT using `search_and_remove_and_yield()` here?"

Comment (mengxiang0811) [reply lib/librte_hash/rte_cuckoo_hash.c]: Thanks for the clarification!

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

File: lib/librte_hash/rte_cuckoo_hash.c [hunk 4]
```diff
@@ -1339,6 +1339,215 @@ rte_hash_lookup_with_hash(const struct rte_hash *h,
...
+		cnt_b = __atomic_load_n(h->tbl_chng_cnt,
+				__ATOMIC_ACQUIRE);
+
+		/* Check if key is in primary location */
+		bkt = &h->buckets[prim_bucket_idx];
+		yield_func(bkt, arg);
+		ret = search_one_bucket_and_yield_lf(h, key, short_sig, data,
+			bkt, yield_func, arg);
+		if (ret != -1) {
+			__hash_rw_reader_unlock(h);
```
Comment (AltraMayor) [inline lib/librte_hash/rte_cuckoo_hash.c:1490]: I know that you copied the line above from DPDK, but it seems to be a bug in DPDK here: this function is lock-free and there's no call to `__hash_rw_reader_lock()`. Could you verify if the latest version of DPDK has fixed it? If you confirm this is a bug and it hasn't been fixed in the latest version of DPDK, it's worth reporting.

This comment also applies to the call `__hash_rw_reader_unlock()` below.

**Review by AltraMayor — CHANGES_REQUESTED** [review]
**hash: add yield version of lookup and deletion functions**

**Comment by AltraMayor** [comment]
This pull request has a fatal issue: it leads to deadlock with coroutines. Consider the following sequence of events: coroutine A gets a read lock by calling `rte_hash_lookup_and_yield_with_hash()` and yields to coroutine B. Coroutine B tries to get a write lock by adding or removing new entries to the hash table. Since Coroutine A did not release the read lock, coroutine B will never make progress.

Using the lockless option of hash tables wouldn't help because whenever a write is done, it would invalidate all the work and prefetches that the other coroutines would be waiting on.

The deadlock is not possible in the current pull request of coroutines because locks are acquired and released before the coroutines yield.

Nevertheless, the idea of passing a yield function down to DPDK is a good idea that can be applied to pull request #16 to combine the functions `rte_lpm_prefetch_tbl24_entry()`, `rte_lpm_lookup_step1()`, `rte_lpm_prefetch_tbl8_entry()`, `rte_lpm_lookup_step2()` there into a single, simpler function. Moreover, the deadlock explained above wouldn't take place with `rte_lpm_lookup_and_yield()` because GK blocks do not edit the FIB.

File: lib/librte_hash/rte_cuckoo_hash.c [hunk 5]
```diff
@@ -1339,6 +1339,215 @@ rte_hash_lookup_with_hash(const struct rte_hash *h,
...
+{
+	int i;
+	struct rte_hash_key *k, *keys = h->key_store;
+
+	for (i = 0; i < RTE_HASH_BUCKET_ENTRIES; i++) {
+		if (bkt->sig_current[i] == sig &&
+				bkt->key_idx[i] != EMPTY_SLOT) {
+			k = (struct rte_hash_key *) ((char *)keys +
+					bkt->key_idx[i] * h->key_entry_size);
+			yield_func(k, arg);
```
Comment (mengxiang0811) [inline lib/librte_hash/rte_cuckoo_hash.c:1356]: In this function call, we are yielding while the lock is acquired.

File: lib/librte_hash/rte_cuckoo_hash.c (hunk 4 above)
Comment (mengxiang0811) [reply lib/librte_hash/rte_cuckoo_hash.c:1490]: After taking a look at [the latest version of DPDK](https://git.dpdk.org/dpdk/tree/lib/librte_hash/rte_cuckoo_hash.c#n1264), it doesn't change this call.

Comment (mengxiang0811) [reply lib/librte_hash/rte_cuckoo_hash.c:1490]: ``__hash_rw_reader_lock()`` is called by ``__rte_hash_lookup_with_hash_l()``, ``__rte_hash_lookup_bulk_l()`` and ``rte_hash_iterate()``, so when ``rte_hash_iterate()`` is called by one thread (e.g., ``t1``) which obtained a reader lock and is reading the item with key ``k1``,  another thread calling ``__rte_hash_lookup_with_hash_lf()`` releases a reader lock due to a failure lookup. At the same time, one thread calling either deletion or addition can successfully get the writer lock and write an item with key ``k1`` into the hash table while thread ``t1`` is accessing the item, then collision happens!

Comment (AltraMayor) [reply lib/librte_hash/rte_cuckoo_hash.c:1490]: Thank you for the investigation.

File: lib/librte_hash/rte_cuckoo_hash.c (hunk 5 above)
Comment (AltraMayor) [reply lib/librte_hash/rte_cuckoo_hash.c:1356]: It is **now** okay because we are yielding with a reader lock.

**Comment by AltraMayor** [comment]
I've changed the code of the coroutines, so they never try to acquire a writer lock. Thus, we can use `rte_hash_lookup_and_yield_with_hash()`. But `rte_hash_del_key_and_yield_with_hash()` is still useless. Could you drop `rte_hash_del_key_and_yield_with_hash()` from this patch?

**Comment by mengxiang0811** [comment]
This rebase dropped ``rte_hash_del_key_and_yield_with_hash()`` from this patch.

**Comment by AltraMayor** [comment]
Please update the title of the patch.

**Comment by mengxiang0811** [comment]
Title updated.

**Pull request merged**